ouroboros-network/src/Ouroboros/Network/Subscription/PeerState.hs

```diff
@@ -0,0 +1,559 @@
+  ( SuspendCommand (..)
+  , suspend
+  -- * Error policy GADT
+  , ErrorPolicy (..)
+  , ConnOrAppException (..)
+  , evalErrorPolicy
+  , evalErrorPolicies
+  -- * PeerStates and its operations
+  , PeerState (..)
+  , threadsToCancel
+  , PeerStates (..)
+  , newPeerStatesVar
+  , runSuspendCommand
+  , registerConsumer
+  , unregisterConsumer
+  , registerProducer
+  , unregisterProducer
+
+  -- * Tracing
+  , ErrorPolicyTrace (..)
+  , traceErrorPolicy
+  , WithAddr (..)
+
+  -- * Re-exports
+  , DiffTime
+
+  -- * Auxiliary functions
+  , alterAndLookup
+  ) where
+
+import qualified Control.Concurrent.Async as Async
+import           Control.Exception (Exception, SomeException (..), assert)
+import           Control.Monad.State
+import           Data.Map.Strict (Map)
+import qualified Data.Map as Map
+import           Data.Maybe (mapMaybe)
+import           Data.List.NonEmpty (NonEmpty (..))
+import           Data.Semigroup (sconcat)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Typeable ( Proxy (..)
+                               , (:~:) (..)
+                               , eqT
+                               , gcast
+                               , tyConName
+                               , typeRepTyCon
+                               , typeRep
+                               )
+import           Text.Printf
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+
+import           Data.Semigroup.Action
+
+-- | Semigroup of commands which acts on 'PeerState'.  The @t@ variable might
+-- be initiated to 'DiffTime' or @Time m@.
+--
+-- This semigroup allows to either suspend both consumer and producer or just
+-- the consumer part.
+--
+data SuspendCommand t
+    = SuspendPeer !t !t
+    -- ^ peer is suspend; The first @t@ is the time until which a local
+    -- producer is suspended, the second one is the time until which a local
+    -- consumer is suspended.
+    | SuspendConsumer !t
+    -- ^ suspend local consumer \/ initiator side until @t@ (this mean we are
+    -- not allowing to communicate with the producer \/ responder of a remote
+    -- peer).
+    | Throw
+    -- ^ throw an error from the main thread.
+    deriving (Eq, Ord, Show)
+
+consumerSuspendedUntil :: SuspendCommand t -> Maybe t
+consumerSuspendedUntil (SuspendPeer _ consT)   = Just consT
+consumerSuspendedUntil (SuspendConsumer consT) = Just consT
+consumerSuspendedUntil Throw                   = Nothing
+
+producerSuspendedUntil :: SuspendCommand t -> Maybe t
+producerSuspendedUntil (SuspendPeer prodT _) = Just prodT
+producerSuspendedUntil (SuspendConsumer _) = Nothing
+producerSuspendedUntil Throw               = Nothing
+
+instance Functor SuspendCommand where
+    fmap f (SuspendPeer prodT consT) = SuspendPeer (f prodT) (f consT)
+    fmap f (SuspendConsumer consT) = SuspendConsumer (f consT)
+    fmap _ Throw = Throw
```

Comment:
Sure, if it's a functor then it can be derived, isn't it.  I think one can prove using free theorems that if there exists a functor instance there's only one.